Introduction


In this project, I build an Artificial Neural Network (ANN) to predict customer churn using the Churn Modelling dataset. Customer churn refers to the loss of customers, and predicting churn is crucial for businesses to implement strategies to retain customers. The Churn Modelling dataset contains information on customers of a bank, including their demographics, account information, and whether they have churned.

In [2]:
import pandas as pd
import numpy as np

In [3]:
df=pd.read_csv('/content/Churn_Modelling.csv')

In [4]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [5]:
X=df.iloc[:,3:13]
X.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [7]:
y=df.iloc[:,13]
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

Encoding the categorical data

Using Label encoder to convert the categorical data into numerical data

In [8]:
from sklearn.preprocessing import LabelEncoder

In [9]:
le=LabelEncoder()

In [10]:
X.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [12]:
X["Geography"]=le.fit_transform(X["Geography"])
X.head()

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619          0  Female   42       2       0.00              1   
1          608          2  Female   41       1   83807.86              1   
2          502          0  Female   42       8  159660.80              3   
3          699          0  Female   39       1       0.00              2   
4          850          2  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [14]:
X["Gender"]=le.fit_transform(X["Gender"])
X.head()

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619          0       0   42       2       0.00              1   
1          608          2       0   41       1   83807.86              1   
2          502          0       0   42       8  159660.80              3   
3          699          0       0   39       1       0.00              2   
4          850          2       0   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

To deploy neural network we need to split the data into train test

In [15]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=0)

In [16]:
print(X_train.shape)
print(X_test.shape)

(7500, 10)
(2500, 10)


Feature Scaling

In [20]:
!pip install sklearn


  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


In [23]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train=sc.fit_transform(X_train)
X_test=sc.fit_transform(X_test)

In [24]:
X_train

array([[-0.73550706,  0.31266104, -1.08726059, ..., -1.55362351,
        -1.03446007, -1.64080994],
       [ 1.02442719, -0.89235324, -1.08726059, ...,  0.64365658,
        -1.03446007, -0.07927152],
       [ 0.80829492,  1.51767532, -1.08726059, ...,  0.64365658,
         0.96668786, -0.99684012],
       ...,
       [ 0.90092304, -0.89235324,  0.91974271, ...,  0.64365658,
        -1.03446007,  1.4133552 ],
       [-0.62229491,  1.51767532, -1.08726059, ...,  0.64365658,
         0.96668786,  0.84496184],
       [-0.28265848,  0.31266104, -1.08726059, ...,  0.64365658,
        -1.03446007,  0.32500428]])

In [25]:
X_test

array([[-0.56498376,  0.28826802, -1.12267441, ...,  0.65340695,
         0.98098083,  1.61610909],
       [-1.34320411, -0.93113986, -1.12267441, ...,  0.65340695,
        -1.01938791,  0.50132237],
       [ 0.58131377,  1.50767591, -1.12267441, ...,  0.65340695,
         0.98098083, -0.41916962],
       ...,
       [ 1.99052359, -0.93113986, -1.12267441, ...,  0.65340695,
        -1.01938791, -0.26702178],
       [ 0.29736851,  0.28826802,  0.8907302 , ...,  0.65340695,
        -1.01938791, -0.92676356],
       [ 0.88629202, -0.93113986,  0.8907302 , ..., -1.5304398 ,
        -1.01938791, -0.46790648]])

Creating own ANN model
Listing the steps:
1.Randomly initialize the weights to small numbers close to 0 but not 0

2.inputing the first observation of the dataset in the input layer each feature in one input node.

3.Forward propagation from left to right the neurons are activated in a way that the impact of each neuron activation is limited by the weights.propagate the activation until getting the predicted result

4.Comparing the predicted result with the actual result. Measuring the generated error

5.back propagation: From right to left error is back propagated.Update the weights according to how much they are responsible for the error.Learning rate  tells us how much we update the weight.

6.Repeating the step 1 to 5 and update the weight after each observation(Reinforcement learning)

7.When the whole training data is passed through the ANN  That completes an Epoch.

Importing Keras library and packages

In [27]:
import keras
# for building the neural network layer by layer
from keras.models import Sequential

In [28]:
# to randomly initailize the weights to small numbers close to 0 (But not 0)
from keras.layers import Dense

Intialising the ANN :
There are two ways to initailise the deep learning model
1.defining each layer one by one
2.defing a Graph

Adding the Input layer and 1st hidden layer

In [29]:
classifier=Sequential()

In [30]:
#defining the input layer
classifier.add(Dense(5,kernel_initializer='glorot_uniform',activation='relu',input_dim=10))

In [34]:
classifier.add(Dense(5,kernel_initializer='uniform',activation='sigmoid'))
classifier.add(Dense(1,kernel_initializer='uniform',activation='sigmoid'))

compiling the ANN

In [36]:
classifier.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])


In [38]:
#Training the model with data
classifier.fit(X_train,y_train,batch_size=10,epochs=100)

Epoch 1/100
750/750 [==============================] - 2s 2ms/step - loss: 0.5377 - accuracy: 0.7963
Epoch 2/100
750/750 [==============================] - 2s 2ms/step - loss: 0.4985 - accuracy: 0.7963
Epoch 3/100
750/750 [==============================] - 2s 3ms/step - loss: 0.4920 - accuracy: 0.7963
Epoch 4/100
750/750 [==============================] - 1s 2ms/step - loss: 0.4811 - accuracy: 0.7963
Epoch 5/100
750/750 [==============================] - 1s 2ms/step - loss: 0.4664 - accuracy: 0.7963
Epoch 6/100
750/750 [==============================] - 2s 2ms/step - loss: 0.4533 - accuracy: 0.7963
Epoch 7/100
750/750 [==============================] - 1s 2ms/step - loss: 0.4451 - accuracy: 0.7963
Epoch 8/100
750/750 [==============================] - 1s 2ms/step - loss: 0.4403 - accuracy: 0.7963
Epoch 9/100
750/750 [==============================] - 1s 2ms/step - loss: 0.4378 - accuracy: 0.7963
Epoch 10/100
750/750 [==============================] - 1s 2ms/step - loss: 0.4363 - accura

Predicting the Test result

In [39]:
 y_pred=classifier.predict(X_test)

79/79 [==============================] - 0s 2ms/step


In [40]:
print(y_pred)

[[0.19951251]
 [0.23049165]
 [0.17363526]
 ...
 [0.18797666]
 [0.22752771]
 [0.06297389]]


In [41]:
churn=[]
churn_val=[]
for i in y_pred:
  if i>0.5:
    churn.append("YES")
    churn_val.append(1)
  else:
    churn.append("NO")
    churn_val.append(0)

In [42]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_test,churn_val)
print(cm)
print(accuracy_score(y_test,churn_val))

[[1916   75]
 [ 314  195]]
0.8444


This model when trained on the train data and made test on test data gave accuracy of around 84% In both cases